### PR TITLE
[CARBONDATA-1317] Multiple dictionary files being created in single_pass

### PR DESCRIPTION
**Pull request by kunal642**
Analysis: The tableName -> TableDictionaryGenerator map was not being cleared once the load was complete that is why when a new load comes after dropping table the old columns are also stored in the TableDictionaryGenerators which cause the dictionary files to be more 

Solution: Update the map from tableName to tableId and clear the map after writing dictionary.

### CODE REVIEW

**Comment by asfgit** [comment]
Can one of the admins verify this patch?

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/2713/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/carbondata-pr-spark-1.6/638/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/138/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/143/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/2718/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/carbondata-pr-spark-1.6/643/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/carbondata-pr-spark-1.6/644/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/2719/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/144/

File: core/src/main/java/org/apache/carbondata/core/dictionary/generator/ServerDictionaryGenerator.java [hunk 1]
```diff
@@ -40,21 +42,21 @@
   @Override
   public Integer generateKey(DictionaryMessage value)
       throws DictionaryGenerationException {
-    TableDictionaryGenerator generator = tableMap.get(value.getTableUniqueName());
-    assert generator != null : "Table initialization for generator is not done";
+    TableDictionaryGenerator generator = tableMap.get(value.getTableUniqueId());
+    if (generator == null) {
+      throw new DictionaryGenerationException("Table Not Found");
```
Comment (manishgupta88) [inline core/src/main/java/org/apache/carbondata/core/dictionary/generator/ServerDictionaryGenerator.java]: Change message in the assert statement....do not throw explicit exception

File: core/src/main/java/org/apache/carbondata/core/dictionary/generator/ServerDictionaryGenerator.java [hunk 2]
```diff
@@ -40,21 +42,21 @@
...
-    CarbonTable carbonTable = metadata.getCarbonTable(key.getTableUniqueName());
+  public void initializeGeneratorForTable(DictionaryMessage key, CarbonTable carbonTable) {
     CarbonDimension dimension = carbonTable.getPrimitiveDimensionByName(
             key.getTableUniqueName(), key.getColumnName());
     // initialize TableDictionaryGenerator first
     if (tableMap.get(key.getTableUniqueName()) == null) {
       synchronized (tableMap) {
         if (tableMap.get(key.getTableUniqueName()) == null) {
-          tableMap.put(key.getTableUniqueName(), new TableDictionaryGenerator(dimension));
+          tableMap.put(key.getTableUniqueName(), new TableDictionaryGenerator(dimension, carbonTable));
```
Comment (manishgupta88) [inline core/src/main/java/org/apache/carbondata/core/dictionary/generator/ServerDictionaryGenerator.java]: make unique name as dbName_tableID

File: core/src/main/java/org/apache/carbondata/core/dictionary/generator/ServerDictionaryGenerator.java [hunk 3]
```diff
@@ -64,24 +66,34 @@ public void initializeGeneratorForTable(DictionaryMessage key) {
     }
   }
 
+  public void initializeGeneratorForColumn(DictionaryMessage key) {
+    tableMap.get(key.getTableUniqueId()).updateGenerator(key);
```
Comment (manishgupta88) [inline core/src/main/java/org/apache/carbondata/core/dictionary/generator/ServerDictionaryGenerator.java]: make sure get Method is not returning null from the map else NullPointerException will be thrown

File: core/src/main/java/org/apache/carbondata/core/dictionary/generator/ServerDictionaryGenerator.java [hunk 4]
```diff
@@ -64,24 +66,34 @@ public void initializeGeneratorForTable(DictionaryMessage key) {
...
+  }
+
   public Integer size(DictionaryMessage key) {
-    TableDictionaryGenerator generator = tableMap.get(key.getTableUniqueName());
+    TableDictionaryGenerator generator = tableMap.get(key.getTableUniqueId());
     assert generator != null : "Table intialization for generator is not done";
```
Comment (manishgupta88) [inline core/src/main/java/org/apache/carbondata/core/dictionary/generator/ServerDictionaryGenerator.java]: Change the message to "Table not found" as above assert statement

File: core/src/main/java/org/apache/carbondata/core/dictionary/generator/ServerDictionaryGenerator.java [hunk 5]
```diff
@@ -64,24 +66,34 @@ public void initializeGeneratorForTable(DictionaryMessage key) {
...
   }
 
-  public void writeTableDictionaryData(String tableUniqueName) throws Exception {
-    TableDictionaryGenerator generator = tableMap.get(tableUniqueName);
-    if (generator != null) {
-      generator.writeDictionaryData(tableUniqueName);
-    }
+  public void writeTableDictionaryData(String tableUniqueId, CarbonTable carbonTable) throws Exception {
+      TableDictionaryGenerator generator = tableMap.get(tableUniqueId);
+    boolean tableExists = FileFactory.isFileExist(carbonTable.getMetaDataFilepath() + CarbonCommonConstants.FILE_SEPARATOR + tableUniqueId,
```
Comment (manishgupta88) [inline core/src/main/java/org/apache/carbondata/core/dictionary/generator/ServerDictionaryGenerator.java]: correct the indentation

File: core/src/main/java/org/apache/carbondata/core/dictionary/generator/key/DictionaryMessage.java [hunk 6]
```diff
@@ -55,6 +60,10 @@ public void readData(ByteBuf byteBuf) {
     byteBuf.readBytes(tableBytes);
     tableUniqueName = new String(tableBytes);
 
+    byte[] tableIdBytes = new byte[byteBuf.readInt()];
+    byteBuf.readBytes(tableIdBytes);
+    tableUniqueId = new String(tableIdBytes);
```
Comment (manishgupta88) [inline core/src/main/java/org/apache/carbondata/core/dictionary/generator/key/DictionaryMessage.java]: creating a string object for every message is a costly operation. Read bytes and return the byte array and maintain a map with byte array as key object

File: core/src/main/java/org/apache/carbondata/core/dictionary/server/DictionaryServer.java [hunk 7]
```diff
@@ -146,21 +147,31 @@ public void shutdown() throws Exception {
...
-    dictionaryServerHandler.processMessage(key);
+    dictionaryServerHandler.processMessage(key, carbonTable);
+  }
+
+  public void initializeDictionaryGenerator(CarbonTable carbonTable) throws Exception {
+    DictionaryMessage key = new DictionaryMessage();
+    key.setType(DictionaryMessageType.TABLE_INTIALIZATION);
+    key.setColumnName("dummy");
+    key.setTableUniqueId(carbonTable.getCarbonTableIdentifier().getTableId());
+    key.setTableUniqueName(carbonTable.getTableUniqueName());
```
Comment (manishgupta88) [inline core/src/main/java/org/apache/carbondata/core/dictionary/server/DictionaryServer.java]: check if tableUniqueName is required...I think we should be able to work using tableUniqueId

File: core/src/main/java/org/apache/carbondata/core/dictionary/server/DictionaryServer.java [hunk 8]
```diff
@@ -146,21 +147,31 @@ public void shutdown() throws Exception {
...
   /**
    *  Write Dictionary for one table.
    * @throws Exception
    */
 
-  public void writeTableDictionary(String uniqueTableName) throws Exception {
+  public void writeTableDictionary(CarbonTable carbonTable) throws Exception {
     DictionaryMessage key = new DictionaryMessage();
-    key.setTableUniqueName(uniqueTableName);
+    key.setTableUniqueName(carbonTable.getTableUniqueName());
```
Comment (manishgupta88) [inline core/src/main/java/org/apache/carbondata/core/dictionary/server/DictionaryServer.java]: same comment as above...check if we can remove tableUniqueName

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/CarbonSessionExample.scala [hunk 9]
```diff
@@ -51,7 +51,7 @@ object CarbonSessionExample {
 
     spark.sql("DROP TABLE IF EXISTS carbon_table")
 
-    // Create table
+//     Create table
```
Comment (manishgupta88) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/CarbonSessionExample.scala]: This class need not be modified...remove the class

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/CarbonGlobalDictionaryRDD.scala [hunk 10]
```diff
@@ -442,7 +456,15 @@ class CarbonGlobalDictionaryGenerateRDD(
...
-            sortIndexWriteTask.execute()
+            breakable {
+              if (!FileUtils
+                .validateTableExists(model.dictfolderPath + CarbonCommonConstants.FILE_SEPARATOR +
+                                     model.table.getTableId)) {
+                LOGGER.error(s"Table does not exists: ${ model.table.getTableUniqueName }")
+                break()
+              }
+              sortIndexWriteTask.execute()
+            }
```
Comment (manishgupta88) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/CarbonGlobalDictionaryRDD.scala]: Remove file exist check from the executor code as it is a costly operation to check for file existence through RPC call to namenode for each column multiple times. Instead add this check in driver after global dictionary generation and delete the dictionary files created in case the table has been deleted.

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/util/GlobalDictionaryUtil.scala [hunk 11]
```diff
@@ -867,6 +868,11 @@ object GlobalDictionaryUtil {
         columnSchema,
         false
       )
+      if (!FileUtils
+        .validateTableExists(carbonTablePath.getMetadataDirectoryPath  + CarbonCommonConstants.FILE_SEPARATOR +
+                             tableIdentifier.getTableId)) {
+        throw new Exception(s"Table does not exists: ${ tableIdentifier.getTableUniqueName }")
+      }
```
Comment (manishgupta88) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/util/GlobalDictionaryUtil.scala]: Remove this check from executor and follow the same method to handle as mentioned in above comment

File: integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala [hunk 12]
```diff
@@ -808,10 +808,10 @@ object CarbonDataRDDFactory {
...
         else {
           val newStatusMap = scala.collection.mutable.Map.empty[String, String]
-        if (status.nonEmpty) {
-          status.foreach { eachLoadStatus =>
-            val state = newStatusMap.get(eachLoadStatus._1)
-            state match {
+          if (status.nonEmpty) {
+            status.foreach { eachLoadStatus =>
+              val state = newStatusMap.get(eachLoadStatus._1)
+              state match {
```
Comment (manishgupta88) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala]: indentation change...revert it back to old...this change is not part of this PR

File: integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala [hunk 13]
```diff
@@ -1099,15 +1099,18 @@ object CarbonDataRDDFactory {
...
     result match {
       case Some(server) =>
         try {
           if (writeAll) {
-            server.writeDictionary()
+            server.writeDictionary(carbonLoadModel.getCarbonDataLoadSchema.getCarbonTable)
           }
           else {
-            server.writeTableDictionary(uniqueTableName)
+//            server.writeTableDictionary(carbonLoadModel.getCarbonDataLoadSchema.getCarbonTable)
```
Comment (manishgupta88) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala]: delete commented line

File: integration/spark2/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala [hunk 14]
```diff
@@ -640,7 +640,16 @@ object CarbonDataRDDFactory {
...
+          breakable{
+            if (!FileUtils
+              .validateTableExists(carbonTable.getAbsoluteTableIdentifier.getStorePath +
+                                   CarbonCommonConstants.FILE_SEPARATOR +
+                                   "MetaData" + CarbonCommonConstants.FILE_SEPARATOR +
+                                   carbonTable.getCarbonTableIdentifier.getTableId)) {
+              LOGGER.error(s"Table does not exists: ${carbonTable.getTableUniqueName}")
+              break()
+            }
+          }
```
Comment (manishgupta88) [inline integration/spark2/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala]: This check should be done only at the final commit point of an operation. Like for data load this should be done just before writing the table status file and after global dictionary generation.

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala [hunk 15]
```diff
@@ -358,7 +358,17 @@ case class LoadTable(
       LOGGER.audit(s"Data loading failed. table not found: $dbName.$tableName")
       sys.error(s"Data loading failed. table not found: $dbName.$tableName")
     }
-
+    LOGGER.info("----------"+relation.metaData.carbonTable.getAbsoluteTableIdentifier.getTablePath +
+                CarbonCommonConstants.FILE_SEPARATOR + "/Metadata/" +
+                relation.metaData.carbonTable.getCarbonTableIdentifier.getTableId)
+    FileFactory.createNewFile(
+      relation.metaData.carbonTable.getAbsoluteTableIdentifier.getTablePath +
+      CarbonCommonConstants.FILE_SEPARATOR + "/Metadata/" +
```
Comment (manishgupta88) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala]: get "/Metadata/" from a constant file

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala [hunk 16]
```diff
@@ -358,7 +358,17 @@ case class LoadTable(
       LOGGER.audit(s"Data loading failed. table not found: $dbName.$tableName")
       sys.error(s"Data loading failed. table not found: $dbName.$tableName")
     }
-
+    LOGGER.info("----------"+relation.metaData.carbonTable.getAbsoluteTableIdentifier.getTablePath +
+                CarbonCommonConstants.FILE_SEPARATOR + "/Metadata/" +
+                relation.metaData.carbonTable.getCarbonTableIdentifier.getTableId)
+    FileFactory.createNewFile(
```
Comment (manishgupta88) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala]: Move this code of creating a table id file at the time of table creation

File: processing/src/main/java/org/apache/carbondata/processing/datatypes/PrimitiveDataType.java [hunk 17]
```diff
@@ -141,11 +147,12 @@ public PrimitiveDataType(String name, String parentname, String columnId,
...
-          dictionaryMessage.setType(DictionaryMessageType.TABLE_INTIALIZATION);
+//          dictionaryMessage.setType(DictionaryMessageType.COLUMN_INITIALIZATION);
+          dictionaryMessage.setTableUniqueId(carbonTableIdentifier.getTableId());
           dictionaryMessage.setData("0");
-          if (tableInitialize) {
-            client.getDictionary(dictionaryMessage);
-          }
+//          if (tableInitialize) {
+//            client.getDictionary(dictionaryMessage);
+//          }
```
Comment (manishgupta88) [inline processing/src/main/java/org/apache/carbondata/processing/datatypes/PrimitiveDataType.java]: Delete commented lines

File: processing/src/main/java/org/apache/carbondata/processing/newflow/AbstractDataLoadProcessorStep.java [hunk 18]
```diff
@@ -95,6 +96,18 @@ public AbstractDataLoadProcessorStep(CarbonDataLoadConfiguration configuration,
...
+          configuration.getTableIdentifier().getTablePath() + "/Metadata/" + configuration.getTableIdentifier().getCarbonTableIdentifier().getTableId(), FileFactory.getFileType(
+              configuration.getTableIdentifier().getTablePath() + "/Metadata/" + configuration.getTableIdentifier().getCarbonTableIdentifier().getTableId()))) {
+        LOGGER.error("Table does not exists: " + configuration.getTableIdentifier().getCarbonTableIdentifier().getTableUniqueName());
+        throw new CarbonDataLoadingException("Table does not exists: " + configuration.getTableIdentifier().getCarbonTableIdentifier().getTableUniqueName());
+      }
+    } catch (CarbonDataLoadingException e) {
+      throw e;
+    } catch (IOException e) {
+
+    }
```
Comment (manishgupta88) [inline processing/src/main/java/org/apache/carbondata/processing/newflow/AbstractDataLoadProcessorStep.java]: This code is not required in executor. Handle in driver

File: processing/src/main/java/org/apache/carbondata/processing/newflow/converter/impl/DictionaryFieldConverterImpl.java [hunk 19]
```diff
@@ -80,11 +86,12 @@ public DictionaryFieldConverterImpl(DataField dataField,
...
-      dictionaryMessage.setType(DictionaryMessageType.TABLE_INTIALIZATION);
+//      dictionaryMessage.setType(DictionaryMessageType.COLUMN_INITIALIZATION);
+      dictionaryMessage.setTableUniqueId(carbonTableIdentifier.getTableId());
       dictionaryMessage.setData("0");
-      if (tableInitialize) {
-        client.getDictionary(dictionaryMessage);
-      }
+//      if (tableInitialize) {
+//        client.getDictionary(dictionaryMessage);
+//      }
```
Comment (manishgupta88) [inline processing/src/main/java/org/apache/carbondata/processing/newflow/converter/impl/DictionaryFieldConverterImpl.java]: Remove commented lines

File: processing/src/main/java/org/apache/carbondata/processing/newflow/steps/DataWriterProcessorStepImpl.java [hunk 20]
```diff
@@ -83,10 +87,20 @@ public CarbonFactDataHandlerModel getDataHandlerModel(int partitionId) {
...
+      if (!FileFactory.isFileExist(
+          configuration.getTableIdentifier().getTablePath() + "/Metadata/" + tableIdentifier.getTableId(), FileFactory.getFileType(
+              configuration.getTableIdentifier().getTablePath() + "/Metadata/" + tableIdentifier.getTableId()))) {
+        LOGGER.error("Table does not exists: " + tableIdentifier.getTableUniqueName());
+        return null;
+      }
+    } catch (IOException e) {
+
+    }
+    Iterator<CarbonRowBatch>[] iterators = child.execute();
```
Comment (manishgupta88) [inline processing/src/main/java/org/apache/carbondata/processing/newflow/steps/DataWriterProcessorStepImpl.java]: Remove this check as it is not required in executor

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/2733/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/158/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/carbondata-pr-spark-1.6/661/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/165/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/2741/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/carbondata-pr-spark-1.6/670/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/2747/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/169/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/carbondata-pr-spark-1.6/676/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/3113/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/522/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/3115/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/524/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/3119/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/528/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/3155/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/562/

**Comment by kunal642** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/3156/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/563/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/3157/

**Comment by CarbonDataQA** [comment]
Can one of the admins verify this patch?

**Comment by kunal642** [comment]
@manishgupta88 @gvramana  review this please

**Comment by CarbonDataQA** [comment]
SDV Build Failed with Spark 2.1, Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/306/

**Comment by manishgupta88** [comment]
ok to test

File: core/src/main/java/org/apache/carbondata/core/dictionary/generator/key/DictionaryMessage.java [hunk 21]
```diff
@@ -112,20 +110,12 @@ private DictionaryMessageType getKeyType(byte type) {
       case 1:
         return DictionaryMessageType.DICT_GENERATION;
       case 2:
-        return DictionaryMessageType.TABLE_INTIALIZATION;
-      case 3:
         return DictionaryMessageType.SIZE;
-      case 4:
-        return DictionaryMessageType.WRITE_DICTIONARY;
```
Comment (manishgupta88) [inline core/src/main/java/org/apache/carbondata/core/dictionary/generator/key/DictionaryMessage.java]: 1. Why WRITE_DICTIONARY is removed...any specific reason?
2. Case 1 and default correspond to same behavior...case 1 statement can be removed or default case should throw exception

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/3429/

**Comment by CarbonDataQA** [comment]
SDV Build Success with Spark 2.1, Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/323/

**Comment by kunal642** [comment]
retest this please

**Comment by kunal642** [comment]
retest this please

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/490/

**Comment by gvramana** [comment]
LGTM

**Comment by gvramana** [comment]
Merged to master: commit: ee5f65f7d3f8aa73eaef13758b06cfffca9f31af
